### PR TITLE
wine: add 64 bit build and module

### PR DESCRIPTION
**Pull request by bendlas**
Introduce a nixos module at `config.programs.wine`, where the release
channel (stable/unstable) and the build (32/64/wow) can be selected.

cc @abbradar @7c6f434c

### CODE REVIEW

**Comment by jagajaga** [comment]
Such things have to be related to `nixpkgs` somehow, not to `nixos` modules.

**Comment by bendlas** [comment]
@jagajaga could you elaborate? In my view, `nixpkgs` provides a sane default config and predefines various interesting overrides. What more functionality from the module should instead live in `nixpkgs`?

**Comment by bendlas** [comment]
Sure, I thought that people would rather override themselves, but those predefined overrides do come in handy. I'll remove the deprecation notice.

**Comment by abbradar** [comment]
I'm somewhat hesitant to have a NixOS module with just overrides; user can just use `winesUnstable.wine32` to install needed version. Modules are usually used when there are other things to set up (e.g. `/etc` files or services).

**Comment by jagajaga** [comment]
@abbradar is right.

**Comment by bendlas** [comment]
I see what you mean. Consider my use case with `wineasio`: Right now, I need to set a `WINEDLLPATH` to use it. I had a mind to set up the `WINEDLLPATH` in the module.

Now that you said it, I'm thinking whether that stuff could be done with pure nixpkgs config.
But that creates a bootstrapping problem between `wine` and `wineasio`. How would I solve that?

 But before throwing out the Idea, please first consider my use case with `wineasio`. That package, I've got it working (pretty much), but right now I need to set a `WINEDLLPATH` for it to be found.

**Comment by abbradar** [comment]
@bendlas Oh, it's good then -- `WINEDLLPATH` is exactly the thing which should be set in the module. Perhaps you would add it to the PR?

**Comment by jagajaga** [comment]
@abbradar why not just `wrapProgram`?

**Comment by abbradar** [comment]
On the other hand though, perhaps you can just use `makeWrapper`.
@jagajaga You were faster ^_^

**Comment by jagajaga** [comment]
@abbradar :fast_forward:

**Comment by bendlas** [comment]
I considered using wrappers, but isn't a derivation full of softlinks to another package kind of ugly?

**Comment by abbradar** [comment]
The nice thing about wrappers is that they would work for non-NixOS guys. Check out how `pidgin-with-plugins` is made, for example.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bendlas** [comment]
Also, there are a couple of murky interactions between the wine binaries. E.g. winecfg

**Comment by bendlas** [comment]
Sorry, misclicked. Yeah, will do that.

**Comment by abbradar** [comment]
Another batch of remarks; sorry for nitpicking!

**Comment by bendlas** [comment]
OK, nm. I already spent a lot of time on figuring out the build, so I won't sink more into it till I've caught up with other stuff. Feel free to send patches, I'll integrate them into the PR branch.

**Comment by bendlas** [comment]
I've eliminated the various wines* packages, instead use normal package attributes and overrides. And I set wine64 and wineWow to 64bit only.

**Comment by bendlas** [comment]
Rebased on current master

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -14345,27 +14345,25 @@ let
...
 
-  # Wine cannot be built in 64-bit; use a 32-bit build instead.
-  wineStable = callPackage_i686 ../misc/emulators/wine/stable.nix {
-    bison = bison2;
+  # Wine defaults to a mixed 64 / 32 build on x86_64 and to pure 32 on x86
+  wine = callPackage ../misc/emulators/wine {
+    wineRelease = config.wine.release or "stable";
+    wineBuild = config.wine.build or (if system == "x86_64-linux" then "wineWow" else "wine32");
   };
+  wineStable = wine.override { wineRelease = "stable"; };
```
Comment (jagajaga) [inline pkgs/top-level/all-packages.nix]: What's the difference between `wineStable` and `wine`?

Comment (bendlas) [inline pkgs/top-level/all-packages.nix]: `wineStable` is fixed to the stable release branch, while the release branch of `wine` is subject to selection through `config.wine`

**Comment by jagajaga** [comment]
Looks good for me. Thank you!

**Comment by abbradar** [comment]
Thanks for the work! I'll try to integrate `wineStaging` later when I have time.

**Comment by abbradar** [comment]
Oh, you are already working on it -- nice ^_^

**Comment by bendlas** [comment]
Yes, https://github.com/NixOS/nixpkgs/issues/7798, but I've stalled on the build issue. Will be some time before I get to it again. If you want I can ping you, before I take another stab, so that we don't duplicate work.

**Comment by abbradar** [comment]
@bendlas Yes, please

**Pull request merged**